mesa/visualization/mpl_space_drawing.py

```diff
@@ -47,69 +47,31 @@
-):
-    """Collect the plotting data for all agents in the space.
-
-    Args:
-        space: The space containing the Agents.
-        agent_portrayal: A callable that is called with the agent and returns a dict
-        color: default color
-        size: default size
-        marker: default marker
-        zorder: default zorder
-
-    agent_portrayal should return a dict, limited to size (size of marker), color (color of marker), zorder (z-order),
-    marker (marker style), alpha, linewidths, and edgecolors
-
-    """
-    arguments = {
-        "s": [],
-        "c": [],
-        "marker": [],
-        "zorder": [],
-        "loc": [],
-        "alpha": [],
-        "edgecolors": [],
-        "linewidths": [],
-    }
+    default_color="tab:blue",
+    default_size=25,
+    default_marker="o",
+    default_zorder: int = 1,
+) -> list[AgentPortrayalStyle]:
+    """Collect agent portrayal data efficiently in one loop."""
+    agent_data_list = []
 
     for agent in space.agents:
-        portray = agent_portrayal(agent)
-        loc = agent.pos
-        if loc is None:
-            loc = agent.cell.coordinate
-
-        arguments["loc"].append(loc)
-        arguments["s"].append(portray.pop("size", size))
-        arguments["c"].append(portray.pop("color", color))
-        arguments["marker"].append(portray.pop("marker", marker))
-        arguments["zorder"].append(portray.pop("zorder", zorder))
-
-        for entry in ["alpha", "edgecolors", "linewidths"]:
-            with contextlib.suppress(KeyError):
-                arguments[entry].append(portray.pop(entry))
-
-        if len(portray) > 0:
-            ignored_fields = list(portray.keys())
-            msg = ", ".join(ignored_fields)
-            warnings.warn(
-                f"the following fields are not used in agent portrayal and thus ignored: {msg}.",
-                stacklevel=2,
+        portrayal_dict = agent_portrayal(agent) or {}
+
+        agent_data_list.append(
+            AgentPortrayalStyle(
+                color=portrayal_dict.get("color", default_color),
+                size=portrayal_dict.get("size", default_size),
+                marker=portrayal_dict.get("marker", default_marker),
+                zorder=portrayal_dict.get("zorder", default_zorder),
+                alpha=portrayal_dict.get("alpha", 1.0),
+                linewidths=portrayal_dict.get("linewidths", 1.0),
+                edgecolors=portrayal_dict.get("edgecolors", "black"),
+                loc=agent.pos if agent.pos is not None else (0, 0),  # Store location
             )
+        )
 
-    data = {
-        k: (np.asarray(v, dtype=object) if k == "marker" else np.asarray(v))
-        for k, v in arguments.items()
-    }
-    # ensures that the tuples in marker dont get converted by numpy to an array resulting in a 2D array
-    arr = np.empty(len(arguments["marker"]), dtype=object)
-    arr[:] = arguments["marker"]
-    data["marker"] = arr
-    return data
+    return agent_data_list
```

Comment:
not sure but it might be better to turn all agent data into numpy arrays already here so we avoid needless iterations later.